Core/Transaction_Handler.php

```diff
@@ -0,0 +1,215 @@
+    {
+        switch ($sActionTitle) {
+            case 'cancel':
+                return $oPayment->getCancelTransaction();
+            case 'pay':
+                return $oPayment->getCaptureTransaction();
+        }
+
+        return null;
+    }
+
+    /**
+     * @param SuccessResponse $oResponse
+     *
+     * @return null
+     */
+    private function _handleNotificationSuccess($oResponse)
+    {
+        $this->_oLogger->debug('Notification response: ' . $oResponse->getRawData());
+        $aData = $oResponse->getData();
+
+        $sOrderId = $aData['order-number'];
+        $oOrder = oxNew(Order::class);
+        if (!$oOrder->load($sOrderId)) {
+            $this->_oLogger->error('OrderId not valid: ' . $sOrderId);
+            return;
+        }
+
+        $sPaymentMethod = $oOrder->oxorder__oxpaymenttype->value;
+        $oPayment = oxNew(Payment::class);
+        $oPayment->load($sPaymentMethod);
+
+        $oTransaction = oxNew(Transaction::class);
+        $oTransaction->wdoxidee_ordertransactions__ordernumber = new Field($oOrder->oxorder__oxordernr->value);
+        $oTransaction->wdoxidee_ordertransactions__orderid = new Field($oOrder->oxorder__oxid->value);
+        $oTransaction->wdoxidee_ordertransactions__transactionid = new Field($oResponse->getTransactionId());
+        $oTransaction->wdoxidee_ordertransactions__parenttransactionid
+            = new Field($oResponse->getParentTransactionId());
+        $oTransaction->wdoxidee_ordertransactions__requestid = new Field($oResponse->getRequestId());
+        $oTransaction->wdoxidee_ordertransactions__action
+            = new Field($oPayment->oxpayments__wdoxidee_transactionaction->value);
+        $oTransaction->wdoxidee_ordertransactions__type = new Field($oResponse->getTransactionType());
+        $oTransaction->wdoxidee_ordertransactions__state = new Field($aData['transaction-state']);
+        $oTransaction->wdoxidee_ordertransactions__amount = new Field($oResponse->getRequestedAmount()->getValue());
+        $oTransaction->wdoxidee_ordertransactions__currency
+            = new Field($oResponse->getRequestedAmount()->getCurrency());
+        $oTransaction->wdoxidee_ordertransactions__responsexml = new Field(base64_encode($oResponse->getRawData()));
+        $oTransaction->wdoxidee_ordertransactions__date = new Field($aData['completion-time-stamp']);
+        $oTransaction->save();
+    }
+
+    /**
+     * @param Transaction $oTransaction
+     *
+     * @return array|Payment_Method
+     */
+    private function _getPaymentMethod(Transaction $oTransaction)
+    {
+        try {
+            return Payment_Method_Factory::create($oTransaction->getPaymentType());
+        } catch (Exception $oExc) {
+            $this->_oLogger->error("Error canceling transaction", [$oExc]);
+            return self::_error($oExc->getMessage());
+        }
+    }
+
+    /**
+     * @return array
+     */
+    private static function _success()
```

Comment:
Why is this (and `_error`) static?

Reply:
because it is just a helper to create the return values.